src/Analysis/Engine/Test/LineFormatterTests.cs

```diff
@@ -449,38 +451,22 @@ public class LineFormatterTests {
         /// <param name="line">The line number to request to be formatted.</param>
         /// <param name="languageVersion">Python language version to format.</param>
         /// <param name="editStart">Where the edit should begin (i.e. when whitespace or a multi-line string begins a line).</param>
-        public static void AssertSingleLineFormat(string text, string expected = null, int line = 1, PythonLanguageVersion languageVersion = PythonLanguageVersion.V37, int editStart = 1) {
-            if (text == null) {
-                throw new ArgumentNullException(nameof(text));
-            }
-
-            if (expected == null) {
-                expected = text;
-            }
+        public static void AssertSingleLineFormat(string text, string expected, int line = 0, PythonLanguageVersion languageVersion = PythonLanguageVersion.V37, int editStart = 0) {
```

Comment:
Well, the `Range` and `Position` are zero-based, as well as the rest of LSP. We can split expectation from input, even provide a real range instead of calculating it from the length of input string (which is kinda a place for false positives).

Reply:
I know that they are zero-based under the hood, but `SourceLocation` and `SourceSpan` aren't, and that's what is used for producing human-readable strings which are printed both in the debugger and by FluentAssertions. I think it might make it more difficult to debug failing tests or write new tests when the error messages and debugger say something than you have to write in the test cases themselves.

The tests could be rewritten to provide explicit ranges to replace, but the line formatter is implemented to always replace from the first non-whitespace character (except in the case of multiline strings) until the end of the line. That's why I wrote the tests to use the input string to produce the replaced range.

Reply:
I've fixed the `ToString` methods for `Position` and `Range`, thanks for pointing to it. As for the `SourceLocation` and `SourceSpan`, they aren't needed for the formatter to work. I don't know why they are here at the first place, editors usually operate over continuous index and plain line information (which is still required with `SourceLocation` and `SourceSpan`, so no benefit here as weel), and I don't understand why they are 1-based. That code is very old - those types were in PTVS code base since 2013.

> Line formatter is implemented to always replace from the first non-whitespace character (except in the case of multiline strings) until the end of the line

`EndOfLineCol` has been ignoring the type of line break, so for `\r\n` it has been always cutting the `\r`. This wasn't caught by any test because of the assumption about the line break, so bug in the test logic has been hiding the bug in the product.

Reply:
Thanks for fixing the line ending thing. I also tested these cases in VS Code, so I'm surprised I didn't noticed them. I guess that VS Code is more lenient with how it breaks lines with inconsistent cases (probably choosing to ignore `\r` since `\n` is plenty). `\r\n` line endings aren't something I usually remember, though I tried to make sure to not ignore them by using the right string splitting code when needed.

In the same vein, I wonder how files with a BOM work with the line formatter. I think it should work fine since it starts at the first token's span, but it might be something to test.

Reply:
We shouldn't care about BOM on the level of formatter, it should be handled by the reader via proper encoding. VSCode also sends/receives ranges without BOM bytes offset.

Reply:
Makes sense.